private/pkg/oauth2/client.go

```diff
@@ -0,0 +1,181 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// Client is an OAuth 2.0 client that can register a device, authorize a device,
+// and poll for the device access token.
+type Client struct {
+	baseURL string
+	client  *http.Client
+}
+
+// NewClient returns a new Client with the given base URL and HTTP client.
+func NewClient(baseURL string, client *http.Client) *Client {
+	return &Client{
+		baseURL: baseURL,
+		client:  client,
+	}
+}
+
+// RegisterDevice registers a new device with the authorization server.
+func (c *Client) RegisterDevice(ctx context.Context, args *DeviceRegistrationRequest) (*DeviceRegistrationResponse, error) {
+	input, err := json.Marshal(args)
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequestWithContext(ctx, "POST", c.baseURL+DeviceRegistrationPath, bytes.NewBuffer(input))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+
+	rsp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer rsp.Body.Close()
+
+	payload := &struct {
+		Error
+		DeviceRegistrationResponse
+	}{}
+	if err := c.parseJSONResponse(rsp, payload); err != nil {
+		return nil, fmt.Errorf("oauth2: invalid response: %w", err)
+	}
+	if payload.ErrorCode != "" {
+		return nil, &payload.Error
+	}
+	if code := rsp.StatusCode; code != http.StatusOK {
+		return nil, fmt.Errorf("oauth2: invalid status: %v", code)
+	}
+	return &payload.DeviceRegistrationResponse, nil
+}
+
+// AuthorizeDevice authorizes a device with the authorization server. The authorization server
+// will return a device code and a user code that the user must use to authorize the device.
+func (c *Client) AuthorizeDevice(ctx context.Context, args *DeviceAuthorizationRequest) (*DeviceAuthorizationResponse, error) {
+	req, err := http.NewRequestWithContext(ctx, "POST", c.baseURL+DeviceAuthorizationPath, strings.NewReader(args.ToValues().Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	rsp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer rsp.Body.Close()
+
+	payload := &struct {
+		Error
+		DeviceAuthorizationResponse
+	}{}
+	if err := c.parseJSONResponse(rsp, payload); err != nil {
+		return nil, fmt.Errorf("oauth2: invalid response: %w", err)
+	}
+	if payload.ErrorCode != "" {
+		return nil, &payload.Error
+	}
+	if code := rsp.StatusCode; code != http.StatusOK {
+		return nil, fmt.Errorf("oauth2: invalid status: %v", code)
+	}
+	return &payload.DeviceAuthorizationResponse, nil
+}
+
+// AccessDeviceToken polls the authorization server for the device access token. The interval
+// parameter specifies the polling interval in seconds.
+func (c *Client) AccessDeviceToken(ctx context.Context, interval int, args *DeviceAccessTokenRequest) (*DeviceAccessTokenResponse, error) {
```

Comment:
Do you know if this `ctx` has a deadline when called? ~Should we have one to avoid this running indefinitely?~

Unrelated, but looking forward to go1.23 timer behavior, https://tip.golang.org/doc/go1.23#timer-changes

Reply:
It will be cancellable, but the context is tied tot the cli close too.